src/core/device.c

```diff
@@ -132,6 +291,23 @@ static int device_load(Unit *u) {
                         log_unit_debug_errno(u, r, "Failed to unescape name: %m");
         }
 
+        if (!verify_io_cost_supported(d)) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "IO cost is not supported");
+                return 0;
+        }
+
+        root_slice_unit = manager_get_unit(u->manager, SPECIAL_ROOT_SLICE);
+        if (!root_slice_unit)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Root slice does not exist");
+
+        r = root_cgroup_apply_block_device_io_cost_qos(root_slice_unit, d);
+        if (r < 0)
+                return r;
+
+        r = root_cgroup_apply_block_device_io_cost_model(root_slice_unit, d);
+        if (r < 0)
+                return r;
```

Comment:
this is not the right place. loading happens when a .device unit is referenced already. which might be when the device shows up, but can be during earliest boot in case some regular service has `After=dev-sda1.device` set or so.

maybe device_set_state() or so, conditionalizing it so that whenever it enters DEVICE_PLUGGED state